draft-ietf-tls-esni.md

```diff
@@ -306,6 +310,81 @@ servers to rotate the keys often and improve forward secrecy.
 Note that the length of this structure MUST NOT exceed 2^16 - 1, as the
 RDLENGTH is only 16 bits {{RFC1035}}.
 
+## Encrypted SNI DNS Resolution {#esni-resolution}
+
+This section describes a client ESNI resolution algorithm using a new "address_set"
+extension described below. Future specifications may introduce new extensions
+and corresponding resolution algorithms.
+
+### Address Set Extension
+
+ESNIKeys records MAY indicate a specific IP address(es) for the host(s) in possession
+of the ESNI private key via the following mandatory "address_set" ESNIKeys extension:
+
+~~~
+    enum {
+        address_set(0x1001), (65535)
+    } ExtensionType;
+~~~
+
+The body of this extension is encoded using the following structure.
+
+~~~~
+    enum {
+        address_v4(4),
+        address_v6(6),
+    } AddressType;
+
+    struct {
+        AddressType address_type;
+        select (address_type) {
+            case address_v4: {
+                opaque ipv4Address[4];
+            }
+            case address_v6: {
+                opaque ipv6Address[16];
+            }
+        }
+    } Address;
+
+    struct {
+        Address address_set<1..2^16-1>;
+    } AddressSet;
+~~~~
+
+address_set
+: A set of Address structures containing IPv4 or IPv6 addresses
+to hosts which have the corresponding private ESNI key.
+
+### Resolution Algorithm
+
+Clients obtain ESNI records by querying the DNS for ESNI-enabled server domains.
+In cases where the domain of the A or AAAA records being resolved do not match the
+SNI Server Name, such as when {{!RFC7838}} is being used, the alternate domain should
+be used for querying the ESNI TXT record. (See Section 2.3 of {{!RFC7838}} for more details.)
+
+Clients SHOULD initiate ESNI queries in parallel alongside normal A or AAAA queries to 
+obtain address information in a timely manner in the event that ESNI is available.
+The following algorithm describes a procedure by which clients can process ESNIKeys
+responses as they arrive to produce addresses for ESNI-capable hosts.
+
+~~~
+1. If an ESNIKeys response with an "address_set" extension arrives before an A or 
+AAAA response, clients SHOULD initiate TLS with ESNI to the provided address(es).
+
+2. If an A or AAAA response arrives before the ESNIKeys response, clients SHOULD wait up
+to CD milliseconds before initiating TLS to either address. (Clients may begin
+TCP connections in this time. QUIC connections should wait.) If an ESNIKeys
+response with an "address_set" extension arrives in this time, clients SHOULD 
+initiate TLS with ESNI to the provided address(es). If an ESNIKeys response 
```

Comment:
So in this design, ESNIKeys containing address(es) causes the A/AAAA results to be entirely ignored?

Reply:
Yep!